File: programs in python/link-list-cycle.py
from typing import Optional
from typing import ListNode

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        s = set()

        while head is not None:
            if head in s:
                return True 
            
            s.add(head)
            head = head.next
        
        return False

File: programs in csharp/link-list-cycle.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        var s = new HashSet<ListNode>();

        while(head != null){
            if(s.Contains(head)){
                return true;
            }
            s.Add(head);
            head = head.next;
        }

        return false; 
    }
}